Add Android Request Interception

jkmassel opened the pull request:
This initial implementation allows us to inject authentication at runtime if needed.

You can test it by making the demo app activity conform to `GutenbergRequestInterceptor`, then adding:
```

    override fun interceptRequest(request: WebResourceRequest): WebResourceRequest {

        if (request.url.host.isNullOrEmpty()) {
            return request
        }

        if (!request.url.host!!.contains("wordpress.com")) {
            return request
        }

        request.requestHeaders["Authorization"] =
            "Bearer [redacted – replace this string]"

        return request
    }
```

Once you've added a bearer token, copy the URL to an image on a private blog, and insert a new image block with that URL. The editor will now display it.

jkmassel committed:
Add Android Request Interception

dcalhoun reviewed (commented):
I attempted testing this quickly by rebasing it atop https://github.com/wordpress-mobile/GutenbergKit/pull/29 and using https://github.com/wordpress-mobile/WordPress-Android/pull/21223 to build WordPress-Android with these changes. 

When testing it this way—and without conforming to `GutenbergRequestInterceptor`—I experience CORS errors for all requests for the site's REST API. Below are screenshots of the network failures and one of the requests with and without these changes applied. 

<details><summary>Screenshots</summary>
<p>

| Request Failures | Request Example |
| - | - |
| <img width="1840" alt="cors-failures" src="https://github.com/user-attachments/assets/1bf460a6-d52e-4905-af7b-008f78033524"> | <img width="1698" alt="Screenshot 2024-10-17 at 12 33 02" src="https://github.com/user-attachments/assets/73f3238c-50a8-49a6-893b-8ee7af857c1f"> |

</p>
</details> 

I didn't spend time debugging it aside from inspecting the requests. Maybe some of the headers are erroneously discarded by the interception logic? WDYT? 

Hopefully https://github.com/wordpress-mobile/GutenbergKit/pull/29 and https://github.com/wordpress-mobile/WordPress-Android/pull/21223 merge later today, which will simplify further testing. **Update:** The referenced pull requests are now merged.

dcalhoun committed:
refactor: Remove redundant semi-colon

Address the lint warning.

dcalhoun commented:
Friendly note: I rebased this atop the latest `trunk` branch to resolve merge conflicts to make testing easier.

dcalhoun committed:
fix: Rely upon the WebView for completing non-intercepted requests (#34)

* fix: Rely upon the WebView for completing non-intercepted requests

Previously, a manual request was created for _all_ requests. This was
unnecessary and resulted in header values being unexpectedly discarded.

* refactor: Remove unused imports

* refactor: Remove redundant semicolon

* refactor: Remove unused import

* refactor: Remove unused view parameter

This was added to mirror the `shouldInterceptRequest` method signature,
but is unnecessary complexity.

* refactor: Separate check and handle request methods

Mirror the iOS method structure.

dcalhoun committed:
refactor: Remove unused dependencies

The networking logic that required these dependencies was hoisted to the
host app.

dcalhoun reviewed (approved):
Changes look sound, testing plan succeeds for me. 🚀 

Noting that I removed (69b8e5417607a68087fdd8c45a7563e955466ae2) the previously added `okhttp` dependencies from this change set, as the networking logic was hoisted to the host app.
